pkgs/development/haskell-modules/generic-builder.nix

```diff
@@ -477,13 +516,47 @@ stdenv.mkDerivation ({
     # TODO: fetch the self from the fixpoint instead
     haddockDir = self: if doHaddock then "${docdir self.doc}/html" else null;
 
-    env = shellFor {
-      packages = p: [ drv ];
-      inherit shellHook;
-    };
-
+    envFunc = { withHoogle ? false, ... } @ args:
+      let
+        name = "ghc-shell-for-${drv.name}";
+
+        withPackages = if withHoogle then ghcWithHoogle else ghcWithPackages;
+
+        ghcEnvForBuild = buildHaskellPackages.ghcWithPackages (_: setupHaskellDepends);
+        ghcEnv = withPackages (_:
+          otherBuildInputsHaskell ++
+          propagatedBuildInputs ++
+          stdenv.lib.optionals (!isCross) setupHaskellDepends);
```

Comment:
Why are the `setupHaskellDepends` only needed if we are not cross compiling?

Reply:
Addressed in newest changes